codec/src/main/java/io/netty/handler/codec/compression/BrotliCompressor.java

```diff
@@ -0,0 +1,112 @@
+/*
+ * Copyright 2021 The Netty Project
+ *
+ * The Netty Project licenses this file to you under the Apache License,
+ * version 2.0 (the "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at:
+ *
+ *   https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+package io.netty.handler.codec.compression;
+
+import com.aayushatharva.brotli4j.encoder.Encoder;
+import io.netty.buffer.ByteBuf;
+import io.netty.buffer.ByteBufAllocator;
+import io.netty.buffer.ByteBufUtil;
+import io.netty.buffer.Unpooled;
+import io.netty.util.internal.ObjectUtil;
+
+import java.io.IOException;
+import java.util.Objects;
+import java.util.function.Supplier;
+
+/**
+ * Compress a {@link ByteBuf} with the brotli format.
+ *
+ * See <a href="https://github.com/google/brotli">brotli</a>.
+ */
+public final class BrotliCompressor implements Compressor {
+    private final Encoder.Parameters parameters;
+    private boolean closed;
+
+    /**
+     * Create a new {@link BrotliCompressor} Instance
+     *
+     * @param parameters {@link Encoder.Parameters} Instance
+     */
+    private BrotliCompressor(Encoder.Parameters parameters) {
+        this.parameters = ObjectUtil.checkNotNull(parameters, "Parameters");
+    }
+
+    /**
+     * Create a new {@link BrotliCompressor} factory
+     *
+     * @param parameters {@link Encoder.Parameters} Instance
+     * @return the factory.
+     */
+    public static Supplier<BrotliCompressor> newFactory(Encoder.Parameters parameters) {
```

Comment:
because we want to create these lazy in the `CompressionHandler` (when the `handlerAdded(...)` is called. Otherwise it is impossible to ensure the `close()` method of the `Compressor` is called and so resources can be freed if needed. The same is true for all the `Decompressor`s.

Reply:
I think by directly providing the suppliers its easer to construct the `CompressionHandler` / `DecompressionHandler` and still fail fast on invalid parameters